frontendGraph/src/main/java/com/virtuslab/gitmachete/frontend/graph/repository/RepositoryGraph.java

```diff
@@ -83,18 +78,46 @@ public Integer getNodeIndex(int nodeId) {
     @SuppressWarnings("index:argument.type.incompatible")
     IGraphElement currentElement = elements.get(nodeIndex);
 
-    if (filter.downNormal && nodeIndex < elements.size() - 1) {
-      currentElement.getDownElementIndexes().stream()
-          .map(i -> GraphEdge.createNormalEdge(nodeIndex, i, GraphEdgeType.USUAL)).forEach(adjacentEdges::add);
-    }
-
     if (filter.upNormal && nodeIndex > 0) {
       int upIndex = currentElement.getUpElementIndex();
       if (upIndex >= 0) {
         adjacentEdges.add(GraphEdge.createNormalEdge(nodeIndex, upIndex, GraphEdgeType.USUAL));
       }
     }
 
+    if (filter.downNormal && nodeIndex < elements.size() - 1) {
+      Integer downElementIndex = currentElement.getDownElementIndex();
+      if (downElementIndex != null) {
+        adjacentEdges.add(GraphEdge.createNormalEdge(nodeIndex, downElementIndex, GraphEdgeType.USUAL));
+      }
+    }
+
     return adjacentEdges;
   }
+
+  /**
+   * Visible edges are the edges that are visible in a row but are NOT directly connected
+   * to the node (branch/commit element) of this row. See {@link RepositoryGraph#getAdjacentEdges} for more details.
+   *
+   * @return list of visible edges in a given node index
+   */
+  @SuppressWarnings("index:argument.type.incompatible")
+  public List<Tuple2<GraphEdge, Integer>> getVisibleEdgesWithPositions(int nodeIndex) {
+    assert nodeIndex >= 0 && nodeIndex < nodesCount() : "Bad nodeIndex: " + nodeIndex;
```

Comment:
The stacktrace I posted implies that `nodesCount() != positionsOfVisibleEdges.size()` :/

Reply:
Should be fine now